### PR TITLE
Update Go to 1.19

### PR DESCRIPTION
**Pull request by tklauser**
Release blog: https://go.dev/blog/go1.19
Release notes: https://go.dev/doc/go1.19

### CODE REVIEW

File: plugins/cilium-cni/main.go [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 // Ensure build fails on versions of Go that are not supported by Cilium.
 // This build tag should be kept in sync with the version specified in go.mod.
-//go:build go1.18
+//go:build go1.19
```
Comment (sayboras) [inline plugins/cilium-cni/main.go:6]: any reason to rename this file to main.go ?

**Review by sayboras — APPROVED** [review]
:rocket: 

Just a question for my understanding, we will update dev VM and CI jenkins soon as well right ?

File: plugins/cilium-cni/main.go (hunk 1 above)
Comment (tklauser) [reply plugins/cilium-cni/main.go:6]: Sorry, I forgot to mention this in the commit message. It's so this bit of `update-go-version` works for `plugins/cilium-cni` as well:

https://github.com/cilium/cilium/blob/09a68d6a5f5db01f8b73debce181f79d798fb8f2/Makefile#L615-L621

All other binaries' `main` funcs are already in `main.go` files, but this was probably missed.

Comment (sayboras) [reply plugins/cilium-cni/main.go:6]: oh right, all good :100:

**Comment by tklauser** [comment]
> Just a question for my understanding, we will update dev VM and CI jenkins soon as well right ?

Yes, will send follow-up PRs to update Go in the dev/CI VMs as well.

**Review by nbusseneau — APPROVED** [review]

File: operator/pkg/ciliumendpointslice/manager.go [hunk 2]
```diff
@@ -192,11 +192,12 @@ func uniqueCESliceName(desiredCESs *CESToCEPMapping) string {
 }
 
 // This function create a new ces and capacity to hold maximum ceps in a CES.
-//  called on 2 different scenarios.
-// 1) During runtime, when ces manager decides to create a new ces, it calls
-//    with an empty name, it generates a random unique name and assign it to the CES.
-// 2) During operator warm boot [after crash or software upgrade], slicing manager
-//    creates a CES, by passing unique name.
+//
+//		called on 2 different scenarios.
```
Comment (qmonnet) [inline operator/pkg/ciliumendpointslice/manager.go]: This one might need a small adjustment, it should be aligned like the sentence above and not like the items below?

File: operator/pkg/ingress/ingress.go [hunk 3]
```diff
@@ -52,11 +52,11 @@ type ingressServiceUpdatedEvent struct {
...
 //   - Manage related child resources for this Ingress
-//   	- Service
-//      - Endpoint
-//      - CiliumEnvoyConfig
+//   - Service
+//   - Endpoint
+//   - CiliumEnvoyConfig
 //   - Manage synced TLS secrets in given namespace
-//		- TLS secrets
+//   - TLS secrets
```
Comment (qmonnet) [inline operator/pkg/ingress/ingress.go]: The different indent level was probably on purpose, and seems lost here?

File: pkg/allocator/allocator.go [hunk 4]
```diff
@@ -49,24 +49,25 @@ const (
...
+//     locally (no Backend interactions)
+//  2. Check local cache updated by watcher, if...
 //
 // ... match found:
 // 2.1 Create a new slave key. This operation is potentially racy as the master
-//     key can be removed in the meantime.
-//       etcd: Create is made conditional on existence of master key
-//       consul: locking
+//
+//	key can be removed in the meantime.
```
Comment (qmonnet) [inline pkg/allocator/allocator.go]: The new line breaks the sentence

File: pkg/azure/types/types.go [hunk 5]
```diff
@@ -44,7 +44,7 @@ type AzureSpec struct {
 
 // AzureStatus is the status of Azure addressing of the node
 //
-// This struct is embedded into v2.CiliumNode
+// # This struct is embedded into v2.CiliumNode
```
Comment (qmonnet) [inline pkg/azure/types/types.go]: Not sure where the `#` comes from :thinking:

File: pkg/crypto/certloader/certloader_test.go [hunk 6]
```diff
@@ -366,19 +366,19 @@ func k8sDataDirName() string {
...
+//   - secret:
+//     items:
+//   - key: ca.crt
+//     path: client-ca.crt
+//   - key: tls.crt
+//     path: server.crt
+//   - key: tls.key
+//     path: server.key
+//     name: hubble-server-certs
+//     optional: true
```
Comment (qmonnet) [inline pkg/crypto/certloader/certloader_test.go]: We lose the YAML syntax here!

File: pkg/datapath/linux/linux_defaults/mark.go [hunk 7]
```diff
@@ -7,17 +7,20 @@ package linux_defaults
 // identify traffic from and to proxies. The mark field is being used in the
 // following way:
 //
-//  1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
+//	1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
```
Comment (qmonnet) [inline pkg/datapath/linux/linux_defaults/mark.go]: The additional spaces throw the indices off with the rest of the diagrams, here and below :(

File: pkg/endpoint/events.go [hunk 8]
```diff
@@ -109,7 +109,7 @@ func (e *Endpoint) PolicyRevisionBumpEvent(rev uint64) {
 	}
 }
 
-/// EndpointNoTrackEvent contains all fields necessary to update the NOTRACK rules.
+// / EndpointNoTrackEvent contains all fields necessary to update the NOTRACK rules.
```
Comment (qmonnet) [inline pkg/endpoint/events.go]: Let's get rid of this `/`?

File: pkg/fqdn/cache.go [hunk 9]
```diff
@@ -803,7 +805,8 @@ func (zombies *DNSZombieMappings) isConnectionAlive(zombie *DNSZombieMapping) bo
 }
 
 // getAliveNames returns all the names that are alive
-//   a name is alive if at least one of the IPs that resolve to it is alive
+//
+//	a name is alive if at least one of the IPs that resolve to it is alive
```
Comment (qmonnet) [inline pkg/fqdn/cache.go]: Should be the same paragraph as above?

File: pkg/fqdn/doc.go [hunk 10]
```diff
@@ -34,8 +34,10 @@
 // |             |   |      DNS       |        |         |     |         |
 // |             |   |  Lookup Cache  +<-------+         |     |         |
 // +------+------+   |                |   DNS  +----+----+     +----+----+
-//        |          +----------------+   Data      ^               ^
-//        v                                         |               |
+//
+//	|          +----------------+   Data      ^               ^
+//	v                                         |               |
+//
```
Comment (qmonnet) [inline pkg/fqdn/doc.go]: Ouch

File: pkg/fqdn/doc.go [hunk 11]
```diff
@@ -53,15 +55,15 @@
...
-//          - matchName: "cilium.io"
+//   - toEndpoints:
+//     toPorts:
+//   - ports:
+//   - port: "53"
+//     protocol: ANY
+//     rules:
+//     dns:
+//   - matchPattern: "*"
+//   - matchName: "cilium.io"
```
Comment (qmonnet) [inline pkg/fqdn/doc.go]: YAML indent lost

File: pkg/fqdn/doc.go [hunk 12]
```diff
@@ -92,9 +94,9 @@
 // They rely on DNS lookup cache information and it must come from the DNS
 // proxy, or via a L7 DNS rule.
 //
-//  - toFQDNs:
-//      - matchName: "my-remote-service.com"
-//      - matchPattern: "bucket.*.my-remote-service.com"
+//   - toFQDNs:
+//   - matchName: "my-remote-service.com"
+//   - matchPattern: "bucket.*.my-remote-service.com"
```
Comment (qmonnet) [inline pkg/fqdn/doc.go]: YAML indent lost

File: pkg/fqdn/doc.go [hunk 13]
```diff
@@ -190,36 +192,47 @@
...
 // +----------+----------+
 // |   Policy toFQDNs    |
 // |      Selectors      |
 // +----------+----------+
-//            |
-//            v
+//
+//	|
+//	v
+//
```
Comment (qmonnet) [inline pkg/fqdn/doc.go]: All diagrams above get broken

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml [hunk 14]
```diff
@@ -883,15 +883,15 @@ spec:
...
-                    within an   individual rule."
+                    \n   - All members of this structure are optional. If omitted
+                    or empty, the     member will have no effect on the rule. \n   -
+                    If multiple members of the structure are specified, then all members
+                    \    must match in order for the rule to take effect. The exception
+                    to this     rule is the ToRequires member; the effects of any
+                    Requires field in any     rule will apply to all other rules as
+                    well. \n   - ToEndpoints, ToCIDR, ToCIDRSet, ToEntities, ToServices
+                    and ToGroups are     mutually exclusive. Only one of these members
+                    may be present within an     individual rule."
```
Comment (qmonnet) [inline pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:894]: Some weird slashes and space in this paragraph, but I suppose this is auto-generated?

File: pkg/k8s/slim/k8s/apis/labels/selector.go [hunk 15]
```diff
@@ -138,7 +138,8 @@ type Requirement struct {
 // (4) If the operator is Exists or DoesNotExist, the value set must be empty.
 // (5) If the operator is Gt or Lt, the values set must contain only one value, which will be interpreted as an integer.
 // (6) The key is invalid due to its length, or sequence
-//     of characters. See validateLabelKey for more details.
+//
+//	of characters. See validateLabelKey for more details.
```
Comment (qmonnet) [inline pkg/k8s/slim/k8s/apis/labels/selector.go]: This should be part of the item above. Probably `(n)` numbers that need to be turned into `n.`.

File: pkg/k8s/slim/k8s/apis/labels/selector.go [hunk 16]
```diff
@@ -197,13 +198,20 @@ func (r *Requirement) hasValue(value string) bool {
...
+//
 // (4) The operator is DoesNotExist or NotIn and Labels does not have the
-//     Requirement's key.
+//
+//	Requirement's key.
+//
 // (5) The operator is GreaterThanOperator or LessThanOperator, and Labels has
-//     the Requirement's key and the corresponding value satisfies mathematical inequality.
+//
+//	the Requirement's key and the corresponding value satisfies mathematical inequality.
```
Comment (qmonnet) [inline pkg/k8s/slim/k8s/apis/labels/selector.go]: Same as above.

File: pkg/k8s/slim/k8s/apis/labels/selector.go [hunk 17]
```diff
@@ -828,32 +836,33 @@ func (p *Parser) parseExactValue() (sets.String, error) {
...
-//  (5) A requirement with just !KEY requires that the KEY not exist.
 //
+//	(1) Inclusion - " in " - denotes that the KEY exists and is equal to any of the
+//	    VALUEs in its requirement
+//	(2) Exclusion - " notin " - denotes that the KEY is not equal to any
+//	    of the VALUEs in its requirement or does not exist
+//	(3) The empty string is a valid VALUE
+//	(4) A requirement with just a KEY - as in "y" above - denotes that
+//	    the KEY exists and can be any VALUE.
+//	(5) A requirement with just !KEY requires that the KEY not exist.
```
Comment (qmonnet) [inline pkg/k8s/slim/k8s/apis/labels/selector.go]: `n.` as well, also it seems the existing indent has preserved this paragraph from being changed.

File: pkg/kvstore/allocator/allocator.go [hunk 18]
```diff
@@ -28,26 +28,29 @@ var (
...
-//   longer backed by at least one slave key, the garbage collector will
-//   eventually release the master key and return it back to the pool.
+//   - basePath/id/1001 => key1
+//
+//   - basePath/id/1002 => key2
+//
+//     Master keys provide the mapping from ID to key. As long as a master key
+//     for an ID exists, the ID is still in use. However, if a master key is no
+//     longer backed by at least one slave key, the garbage collector will
+//     eventually release the master key and return it back to the pool.
```
Comment (qmonnet) [inline pkg/kvstore/allocator/allocator.go]: The last paragraph should not be indented like the items above?

File: pkg/policy/l4.go [hunk 19]
```diff
@@ -316,10 +316,11 @@ func (l4 *L4Filter) GetPort() uint16 {
...
 
 // ToMapState converts filter into a MapState with two possible values:
-// - Entry with ProxyPort = 0: No proxy redirection is needed for this key
-// - Entry with any other port #: Proxy redirection is required for this key,
-//                                caller must replace the ProxyPort with the actual
-//                                listening port number.
+//   - Entry with ProxyPort = 0: No proxy redirection is needed for this key
+//   - Entry with any other port #: Proxy redirection is required for this key,
+//     caller must replace the ProxyPort with the actual
+//     listening port number.
```
Comment (qmonnet) [inline pkg/policy/l4.go]: We could remove the `-` at the beginning to have a description list and keep the indent clearer

File: pkg/policy/l4_filter_test.go [hunk 20]
```diff
@@ -1555,8 +1555,9 @@ func (ds *PolicyTestSuite) TestMergingWithDifferentEndpointSelectedAllowAllL7(c
 }
 
 // Case 12: allow all at L3 in one rule with restrictions at L7. Determine that
-//          the host should always be allowed. From Host should go to proxy
-//          allow all; other L3 should restrict at L7 in a separate filter.
+//
+//	the host should always be allowed. From Host should go to proxy
+//	allow all; other L3 should restrict at L7 in a separate filter.
```
Comment (qmonnet) [inline pkg/policy/l4_filter_test.go]: Broken paragraph

**Review by qmonnet — CHANGES_REQUESTED** [review]
Thanks a lot! I'm afraid the `gofmt` changes break the comments at a number of locations and will require some prior adjustments, I've commented below on everything I found.

File: operator/pkg/ingress/ingress.go (hunk 3 above)
Comment (tklauser) [reply operator/pkg/ingress/ingress.go]: Unfortunately, it looks like `gofmt` doesn't support nested lists :(

> List items only contain paragraphs, not code blocks or nested lists.

https://go.dev/doc/comment#lists

If rewritten the comment to use a mix of numbered and unnumbered lists as suggested in https://go.dev/doc/comment#mistakes.

File: pkg/azure/types/types.go (hunk 5 above)
Comment (tklauser) [reply pkg/azure/types/types.go]: `gofmt` probably interprets this as the start of a list due to the empty line above :thinking:

File: pkg/crypto/certloader/certloader_test.go (hunk 6 above)
Comment (tklauser) [reply pkg/crypto/certloader/certloader_test.go]: > A code block is a span of indented or blank lines not starting with a bullet list marker or numbered list marker. It is rendered as preformatted text (a &lt;pre&gt; block in HTML).

-- https://go.dev/doc/comment#code

:sob:

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml (hunk 14 above)
Comment (tklauser) [reply pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:894]: Yes, this is auto-generated and contains the respective godoc comment (which was changed in the corresponding Go file this was generated from).

**Comment by tklauser** [comment]
Thanks for the detailed godoc review @qmo. I've tried to resolve all locations where comments were broken by `gofmt`. FWIW, for YAML code block I've had to resort to prefixing them with `---` because otherwise they would always be treated as lists :( see https://github.com/cilium/cilium/pull/20816#discussion_r941122932

**Review by qmonnet — APPROVED** [review]
> Thanks for the detailed godoc review @qmo. I've tried to resolve all locations where comments were broken by `gofmt`.

Thanks!

> FWIW, for YAML code block I've had to resort to prefixing them with `---` because otherwise they would always be treated as lists :( see https://github.com/cilium/cilium/pull/20816#discussion_r941122932

The `---` looks absolutely fine to me.

Nice work, thanks for addressing it all! I can't really exclude missing things because of the number of formatting changes in the PR, but still I did a second quick pass and didn't notice any broken comment this time. Looks all good!

**Comment by tklauser** [comment]
> Nice work, thanks for addressing it all! I can't really exclude missing things because of the number of formatting changes in the PR, but still I did a second quick pass and didn't notice any broken comment this time. Looks all good!

I've also briefly skimmed the godoc changes and didn't notice anything particularly off. In case something still slipped in, it's at least bisectable to the respective commit and the original formatting could be restored and adjusted from git history.

File: api/v1/client/daemon/daemon_client.go [hunk 21]
```diff
@@ -48,7 +48,7 @@ type ClientService interface {
 }
 
 /*
-  GetClusterNodes gets nodes information stored in the cilium agent
```
Comment (gandro) [inline api/v1/client/daemon/daemon_client.go:51]: Almost all `.go` files in `api/v1` are auto-generated. Are these formatting changes due to the builder base image change or did you explicitly gofmt those files? If the latter, then they'll likely get undone if you run `make -C api/v1` again.

Comment (tklauser) [reply api/v1/client/daemon/daemon_client.go:51]: These got `gofmt`ed like that when running `contrib/scripts/check-api-code-gen.sh` (which in turn runs `make -C api/v1`) due to the builder image being update to Go 1.19, it seems. This is also what's run as part of CI in the `generate-api` step of the `lint-go` workflow.

**Review by gandro — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Review by aanm — CHANGES_REQUESTED** [review]
Sebastian brought a good point with regarding the auto-generated files. Perhaps we should run `gofmt` after we generate the files similar that is done with goimports in here https://github.com/cilium/cilium/blob/9e6f2bb9ee6c6811583c42b239aaccb15eb998de/Makefile#L321-L324
here https://github.com/cilium/cilium/blob/9e6f2bb9ee6c6811583c42b239aaccb15eb998de/Makefile#L341 and here https://github.com/cilium/cilium/blob/9e6f2bb9ee6c6811583c42b239aaccb15eb998de/Makefile#L358

**Comment by tklauser** [comment]
> Sebastian brought a good point with regarding the auto-generated files. Perhaps we should run `gofmt` after we generate the files similar that is done with goimports in here

`goimports` will also format the code in the same way `gofmt` does, see https://pkg.go.dev/golang.org/x/tools/cmd/goimports:

> In addition to fixing imports, goimports also formats your code in the same style as gofmt so it can be used as a replacement for your editor's gofmt-on-save hook.

So I think the existing targets should already be sufficient.

**Review by aanm — APPROVED** [review]

**Pull request merged**